pkg/resources/reconciler.go

```diff
@@ -0,0 +1,137 @@
+package resources
+
+import (
+	"context"
+	"fmt"
+	"github.com/integr8ly/integreatly-operator/pkg/apis/integreatly/v1alpha1"
+	"github.com/integr8ly/integreatly-operator/pkg/controller/installation/marketplace"
+	"github.com/integr8ly/integreatly-operator/pkg/controller/installation/products/config"
+	v1alpha12 "github.com/operator-framework/operator-lifecycle-manager/pkg/api/apis/operators/v1alpha1"
+	"github.com/pkg/errors"
+	"github.com/sirupsen/logrus"
+	v1 "k8s.io/api/core/v1"
+	errors2 "k8s.io/apimachinery/pkg/api/errors"
+	v12 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	pkgclient "sigs.k8s.io/controller-runtime/pkg/client"
+	"time"
+)
+
+// This is the base reconciler that all the other reconcilers extend. It handles things like namespace creation, subscription creation etc
+
+type Reconciler struct {
+	mpm marketplace.MarketplaceInterface
+}
+
+func NewReconciler(mpm marketplace.MarketplaceInterface) *Reconciler {
+	return &Reconciler{
+		mpm: mpm,
+	}
+}
+
+func (r *Reconciler) ReconcileNamespace(ctx context.Context, namespace string, inst *v1alpha1.Installation, client pkgclient.Client) (v1alpha1.StatusPhase, error) {
+	ns := &v1.Namespace{
+		ObjectMeta: v12.ObjectMeta{
+			Name: namespace,
+		},
+	}
+	err := client.Get(ctx, pkgclient.ObjectKey{Name: ns.Name}, ns)
+	if err != nil {
+		if !errors2.IsNotFound(err) {
+			return v1alpha1.PhaseFailed, errors.Wrap(err, fmt.Sprintf("could not retrieve namespace: %s", ns.Name))
+		}
+		prepareNS(ns, inst)
+		if err = client.Create(ctx, ns); err != nil {
+			return v1alpha1.PhaseFailed, errors.Wrap(err, fmt.Sprintf("could not create namespace: %s", ns.Name))
+		}
+		return v1alpha1.PhaseNone, nil
+	}
+	fmt.Println(" ReconcileNamespace ", ns)
+	// ns exists so check it is our namespace
+	if !NSIsOwnedBy(ns, inst) && ns.Status.Phase != v1.NamespaceTerminating {
+		return v1alpha1.PhaseFailed, errors.New("existing namespace found with name " + ns.Name + " but it is not owned by the integreatly installation and it isn't being deleted")
+	}
+	if ns.Status.Phase == v1.NamespaceTerminating {
+		logrus.Debugf("namespace %s is terminating, maintaining phase to try again on next reconcile", namespace)
+		return v1alpha1.PhaseAwaitingNS, nil
+	}
+	prepareNS(ns, inst)
+	if err := client.Update(ctx, ns); err != nil {
+		return v1alpha1.PhaseFailed, errors.Wrap(err, "failed to update the ns definition ")
+	}
+	if ns.Status.Phase != v1.NamespaceActive {
+		return v1alpha1.PhaseAwaitingNS, nil
+	}
+	return v1alpha1.PhaseNone, nil
+}
+
+func (r *Reconciler) ReconcileSubscription(ctx context.Context, subscriptionName, namespace string, client pkgclient.Client) (v1alpha1.StatusPhase, error) {
+	logrus.Infof("reconciling subscription %s from channel %s in namespace: %s", subscriptionName, "integreatly", namespace)
```

Comment:
We can replace the hardcoded `"integreatly"` with `marketplace.IntegreatlyChannel`here